mono/metadata/icall-table.h

```diff
@@ -24,4 +24,352 @@ mono_install_icall_table_callbacks (MonoIcallTableCallbacks *cb);
+// Marshaling a "ptr" does nothing -- just pass it on unchanged.
+// Marshaling a "ref" also does nothing at this layer, but
+// creates a handle in  marshal-ilgen.c.
+typedef gint32  *gint32_ptr;
+typedef gint64  *gint64_ptr;
+typedef guint   *guint_ptr;
+typedef guint32 *guint32_ptr;
+typedef guint64 *guint64_ptr;
+typedef gsize *gsize_ptr;
+typedef guchar *guchar_ptr;
+typedef const guchar *const_guchar_ptr;
+typedef gpointer *gpointer_ptr;
+typedef const char *const_char_ptr;
+typedef char *char_ptr;
+typedef char **char_ptr_ptr;
+typedef gunichar2 *gunichar2_ptr;
+typedef const gunichar2 *const_gunichar2_ptr;
+typedef int *int_ptr;
+typedef int **int_ptr_ref;
+typedef guint8 **guint8_ptr_ref;
+typedef GPtrArray *GPtrArray_ptr;
+typedef MonoAssemblyName *MonoAssemblyName_ptr;
+typedef MonoBoolean *MonoBoolean_ptr;
+typedef MonoClass *MonoClass_ptr;
+typedef MonoClassField *MonoClassField_ptr;
+typedef MonoEvent *MonoEvent_ptr;
+typedef MonoImage *MonoImage_ptr;
+typedef MonoMethod *MonoMethod_ptr;
+typedef MonoNativeOverlapped *MonoNativeOverlapped_ptr;
+typedef MonoProperty *MonoProperty_ptr;
+typedef MonoPropertyInfo *MonoPropertyInfo_ref;
+typedef MonoType *MonoType_ptr;
+typedef MonoTypedRef *MonoTypedRef_ptr;
+typedef gunichar2 *mono_bstr;
+typedef const gunichar2 *mono_bstr_const;
+typedef MonoIUnknown *MonoIUnknown_ptr;
+typedef unsigned *unsigned_ptr;
+typedef mono_unichar2 *mono_unichar2_ptr;
+typedef WSABUF *WSABUF_ptr;
+
+typedef char **char_ptr_ref;
+typedef gint32  *gint32_ref;
+typedef gint64  *gint64_ref;
+typedef gpointer *gpointer_ref;
+typedef gsize *gsize_ref;
+typedef guint32 *guint32_ref;
+typedef guint64 *guint64_ref;
+typedef int *int_ref;
+typedef MonoAssemblyName *MonoAssemblyName_ref;
+typedef MonoBoolean *MonoBoolean_ref;
+typedef MonoClassField *MonoClassField_ref;
+typedef MonoEvent *MonoEvent_ref;
+typedef MonoEventInfo *MonoEventInfo_ref;
+typedef MonoMethod *MonoMethod_ref;
+typedef MonoMethodInfo *MonoMethodInfo_ref;
+typedef MonoNativeOverlapped *MonoNativeOverlapped_ref;
+typedef MonoResolveTokenError *MonoResolveTokenError_ref;
+typedef MonoType *MonoType_ref;
+typedef MonoTypedRef *MonoTypedRef_ref;
+typedef MonoW32ProcessInfo *MonoW32ProcessInfo_ref;
+typedef MonoGenericParamInfo *MonoGenericParamInfo_ptr;
+
+// Maybe do this in TYPED_HANDLE_DECL.
+typedef MonoArray MonoArrayOut;
+typedef MonoArray MonoArrayInOut;
+typedef MonoObject MonoObjectOut;
+typedef MonoObject MonoObjectInOut;
+typedef MonoObjectHandle MonoObjectOutHandle;
+typedef MonoObjectHandle MonoObjectInOutHandle;
+typedef MonoArrayHandle MonoArrayOutHandle;
+typedef MonoArrayHandle MonoArrayInOutHandle;
+typedef MonoString MonoStringOut;
+typedef MonoStringHandle MonoStringOutHandle;
+typedef MonoReflectionModule MonoReflectionModuleOut;
+typedef MonoReflectionModuleHandle MonoReflectionModuleOutHandle;
+
+// How the arguments and return value of an icall should be wrapped.
```

Comment:
Maybe too much duplication of comments from marshal-ilgen.c.
In earlier version, this was a comment move, not a comment copy.

Reply:
Then let's not duplicate it.

Reply:
Agreed.